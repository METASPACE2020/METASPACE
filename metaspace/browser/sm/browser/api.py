import io
import functools

import PIL.Image
import numpy as np
from fastapi import FastAPI
from pydantic import BaseModel
from starlette.responses import StreamingResponse
import uvicorn

from sm.browser import utils
from sm.browser.main import preprocess_dataset_peaks, DatasetBrowser

app = FastAPI()


@functools.lru_cache(maxsize=128)
def load_dataset_browser(s3_path: str):
    return DatasetBrowser(s3_path)


class DatasetPreprocessItem(BaseModel):
    s3_path: str


@app.post("/preprocess")
async def preprocess(item: DatasetPreprocessItem):
    preprocess_dataset_peaks(item.s3_path)
    return {"status": "ok"}


class MzSearchItem(BaseModel):
    s3_path: str
    mz: float
    ppm: int = 3


class PngStreamingResponse(StreamingResponse):
    media_type = "image/png"


@app.post("/search", response_class=PngStreamingResponse)
async def perform_search(item: MzSearchItem):
    dataset_browser = load_dataset_browser(item.s3_path)
    mz_lo, mz_hi = utils.mz_ppm_bin(mz=item.mz, ppm=item.ppm)
    rgba_array = dataset_browser.search(mz_lo, mz_hi)

    image = PIL.Image.fromarray((rgba_array * 255).astype(np.uint8), mode="RGBA")
    fp = io.BytesIO()
    image.save(fp, format="PNG")
    fp.seek(0)
    return PngStreamingResponse(fp)


if __name__ == "__main__":
    uvicorn.run(app)
