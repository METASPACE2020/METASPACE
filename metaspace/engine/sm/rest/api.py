import argparse
import json
import logging
from bottle import post, run
from bottle import request as req
from bottle import response as resp

from sm.engine.db import DB
from sm.engine.es_export import ESExporter
from sm.engine.dataset import Dataset
from sm.engine.png_generator import ImageStoreServiceWrapper
from sm.engine.queue import QueuePublisher, SM_ANNOTATE, SM_DS_STATUS, SM_UPDATE
from sm.engine.util import SMConfig
from sm.engine.util import init_loggers
from sm.engine.errors import UnknownDSID, DSIsBusy
from sm.rest.dataset_manager import SMapiDatasetManager, DatasetActionPriority

OK = {
    'status_code': 200,
    'status': 'success'
}

ERR_DS_NOT_EXIST = {
    'status_code': 404,
    'status': 'not_exist'
}

ERR_OBJECT_EXISTS = {
    'status_code': 400,
    'status': 'already_exists'
}

ERR_DS_BUSY = {
    'status_code': 409,
    'status': 'dataset_busy'
}

ERROR = {
    'status_code': 500,
    'status': 'server_error'
}


def _create_db_conn():
    config = SMConfig.get_conf()
    return DB(config['db'])


def _json_params(req):
    b = req.body.getvalue()
    return json.loads(b.decode('utf-8'))


def _create_queue_publisher(qdesc):
    config = SMConfig.get_conf()
    return QueuePublisher(config['rabbitmq'], qdesc, logger)


def _create_dataset_manager(db):
    config = SMConfig.get_conf()
    img_store = ImageStoreServiceWrapper(config['services']['img_service_url'])
    img_store.storage_type = 'fs'
    return SMapiDatasetManager(db=db, es=ESExporter(db), image_store=img_store,
                               annot_queue=_create_queue_publisher(SM_ANNOTATE),
                               update_queue=_create_queue_publisher(SM_UPDATE),
                               status_queue=_create_queue_publisher(SM_DS_STATUS),
                               logger=logger)


@post('/v1/datasets/<ds_id>/add')
@post('/v1/datasets/add')
def add_ds(ds_id=None):
    """
    :param ds_id: string

    Request params: {
        doc {
            name
            input_path
            upload_dt
            metadata
            is_public
            mol_dbs
            adducts
        }
        priority
        force
        del_first
        email
    }
    """
    ds_id = None
    try:
        params = _json_params(req)
        logger.info(f'Received ADD request: {params}')
        doc = params.get('doc', None)
        if not doc:
            msg = 'No input to create a dataset'
            logger.info(msg)
            raise Exception(msg)
        else:
            db = _create_db_conn()
            ds_man = _create_dataset_manager(db)
            if 'ds_id' in params:
                doc['id'] = params['ds_id']
            ds_id = ds_man.add(
                doc=doc,
                del_first=params.get('del_first', False),
                force=params.get('force', False),
                email=params.get('email', None),
                priority=params.get('priority', DatasetActionPriority.DEFAULT))
            db.close()
            return {
                'status': OK['status'],
                'ds_id': ds_id
            }
    except DSIsBusy as e:
        logger.warning(e.message)
        resp.status = ERR_DS_BUSY['status_code']
        return {
            'status': ERR_DS_BUSY['status'],
            'ds_id': e.ds_id
        }
    except Exception as e:
        logger.error(e, exc_info=True)
        resp.status = ERROR['status_code']
        return {
            'status': ERROR['status'],
            'ds_id': ds_id
        }


def sm_modify_dataset(request_name):
    def _modify(handler):
        def _func(ds_id):
            try:
                params = _json_params(req)
                logger.info('Received %s request: %s', request_name, params)
                db = _create_db_conn()
                ds_man = _create_dataset_manager(db)
                handler(ds_man, ds_id, params)

                db.close()
                return {
                    'status': OK['status'],
                    'ds_id': ds_id
                }
            except UnknownDSID as e:
                logger.warning(e.message)
                resp.status = ERR_DS_NOT_EXIST['status_code']
                return {
                    'status': ERR_DS_NOT_EXIST['status'],
                    'ds_id': e.ds_id
                }
            except DSIsBusy as e:
                logger.warning(e.message)
                resp.status = ERR_DS_BUSY['status_code']
                return {
                    'status': ERR_DS_BUSY['status'],
                    'ds_id': e.ds_id
                }
            except Exception as e:
                logger.error(e, exc_info=True)
                resp.status = ERROR['status_code']
                return {
                    'status': ERROR['status'],
                    'ds_id': ds_id
                }
        return _func
    return _modify


@post('/v1/datasets/<ds_id>/update')
@sm_modify_dataset('UPDATE')
def update_ds(ds_man, ds_id, params):
    """
    :param ds_man: rest.SMapiDatasetManager
    :param ds_id: string
    :param params: {
        doc {
            name
            input_path
            upload_dt
            metadata
            config
            is_public
            submitter_id
            group_id
        }
    }
    :return:
    """
    doc = params.get('doc', None)
    if not doc:
        logger.info(f'Nothing to update for "{ds_id}"')
    else:
        force = params.get('force', False)
        priority = params.get('priority', DatasetActionPriority.STANDARD)
        ds_man.update(ds_id=ds_id, doc=doc,
                      force=force, priority=priority)


@post('/v1/datasets/<ds_id>/delete')
@sm_modify_dataset('DELETE')
def delete_ds(ds_man, ds_id, params):
    del_raw = params.get('del_raw', False)
    force = params.get('force', False)
    ds_man.delete(ds_id=ds_id, del_raw_data=del_raw, force=force)


@post('/v1/datasets/<ds_id>/add-optical-image')
@sm_modify_dataset('ADD_OPTICAL_IMAGE')
def add_optical_image(ds_man, ds_id, params):
    img_id = params['url'].split('/')[-1]
    ds_man.add_optical_image(ds_id, img_id, params['transform'])


@post('/v1/datasets/<ds_id>/del-optical-image')
@sm_modify_dataset('DEL_OPTICAL_IMAGE')
def del_optical_image(ds_man, ds_id, params):
    ds_man.del_optical_image(ds_id)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='SM Engine REST API')
    parser.add_argument('--config', dest='config_path', default='conf/config.json', type=str, help='SM config path')
    args = parser.parse_args()
    SMConfig.set_path(args.config_path)

    init_loggers(SMConfig.get_conf()['logs'])
    logger = logging.getLogger(name='api')
    logger.info('Starting SM api')
    run(**SMConfig.get_conf()['bottle'])
