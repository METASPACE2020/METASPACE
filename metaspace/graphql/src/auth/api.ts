import { Express } from 'express';
import {callbackify} from 'util';
import * as Passport from 'passport';
import {Strategy as LocalStrategy} from 'passport-local';
import {Strategy as GoogleStrategy} from 'passport-google-oauth20';
import * as config from 'config';
import {
  createResetPasswordToken,
  createUser,
  DbUser,
  findUserByEmail,
  findUserByGoogleId,
  findUserById,
  setUserPassword,
} from './db';

const configurePassport = (app: Express) => {
  app.use(Passport.initialize());
  app.use(Passport.session());

  Passport.serializeUser<DbUser, string>(callbackify( async (user: DbUser) => user.id));

  Passport.deserializeUser<DbUser | false, string>(callbackify(async (id: string) => {
    return await findUserById(id) || false;
  }));

  app.post('/api_auth/signout', (req, res) => {
    req.logout();
    res.send('OK');
  });
};

const configureLocalAuth = (app: Express) => {
  Passport.use(new LocalStrategy(
    {
      usernameField: 'email',
    },
    callbackify(async (username: string, password: string) => {
      const user = await findUserByEmail(username);
      return user && user.password === password ? user : false;
    })
  ));

  app.post('/api_auth/signin', Passport.authenticate('local', {
    failureRedirect: '/#/account/sign-in',
  }))
};

const configureGoogleAuth = (app: Express) => {
  const conf = config as any;
  if (conf.GOOGLE_CLIENT_ID) {
    Passport.use(new GoogleStrategy(
      {
        clientID: conf.GOOGLE_CLIENT_ID,
        clientSecret: conf.GOOGLE_CLIENT_SECRET,
        callbackURL: conf.GOOGLE_CALLBACK_URL,
      },
      callbackify(async (accessToken: string, refreshToken: string, profile: any) => {
        return await findUserByGoogleId(profile.id)
          || await createUser({
            googleId: profile.id,
            name: profile.displayName,
            email: profile.emails[0].value,
          });
      })
    ));

    app.get('/api_auth/google', Passport.authenticate('google', {
      scope: ['profile', 'email']
    }));

    app.get('/api_auth/google/callback', Passport.authenticate('google', {
      successRedirect: '/#/datasets',
      failureRedirect: '/#/account/sign-in',
    }));
  }
};

const configureCreateAccount = (app: Express) => {
  app.post('/api_auth/createaccount', async (req, res, next) => {
    try {
      const { name, email, password } = req.body;
      const user = await createUser({ name, email, password });
      res.send(true);
    } catch (err) {
      next(err);
    }
  });

  app.get('/api_auth/verifyemail', async (req, res, next) => {
    const {email, token} = req.query;
    // TODO: Verify email
    const user = findUserByEmail(email);
    if (user) {
      req.login(user, (err) => {
        if (err) {
          next(err);
        } else {
          res.redirect('/#/');
        }
      });
    } else {
      res.sendStatus(401); //TODO: Redirect to a user-friendly error message
    }
  });
};

const configureResetPassword = (app: Express) => {
  app.post('/api_auth/sendpasswordresettoken', async (req, res, next) => {
    try {
      const { email } = req.body;
      const token = await createResetPasswordToken(email);
      console.log(`/#/account/reset-password?email=${encodeURIComponent(email)}&token=${encodeURIComponent(token)}`);
      res.send(true);
    } catch (err) {
      next(err);
    }
  });

  app.post('/api_auth/validatepasswordresettoken', async (req, res, next) => {
    try {
      const { email, token } = req.body;
      const user = await findUserByEmail(email);
      // TODO: token expiry, etc.
      if (user && user.resetPasswordToken === token) {
        res.send(true);
      } else {
        res.sendStatus(400);
      }
    } catch (err) {
      next(err);
    }
  });

  app.post('/api_auth/resetpassword', async (req, res, next) => {
    try {
      const { email, token, password } = req.body;
      let user = await findUserByEmail(email);
      // TODO: token expiry, etc.
      if (user && user.resetPasswordToken === token) {
        user = await setUserPassword(user.id, password);
        req.login(user, (err) => {
          if (err) {
            next(err);
          } else {
            res.send(true);
          }
        });
      } else {
        res.sendStatus(400);
      }
    } catch (err) {
      next(err);
    }
  });
};

export const configureAuth = (app: Express) => {
  configurePassport(app);
  configureLocalAuth(app);
  configureGoogleAuth(app);
  // TODO: find a parameter validation middleware
  configureCreateAccount(app);
  configureResetPassword(app);
};
