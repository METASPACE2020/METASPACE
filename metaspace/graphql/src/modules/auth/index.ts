export {configureAuth} from './controller';
export {initOperation} from './operation';
