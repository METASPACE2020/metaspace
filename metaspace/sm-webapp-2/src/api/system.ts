import gql from 'graphql-tag'

export interface SystemHealth {
  canMutate: boolean
  canProcessDatasets: boolean
  message?: string
}

// Always use fetchPolicy: 'cache-first' for this
export const getSystemHealthQuery = gql`
  query GetSystemHealth {
    systemHealth {
      canMutate
      canProcessDatasets
      message
    }
  }
`

// Always use fetchPolicy: 'cache-first' for this
export const getSystemHealthSubscribeToMore = {
  document: gql`
    subscription SystemHealth {
      systemHealthUpdated {
        canMutate
        canProcessDatasets
        message
      }
    }
  `,
  updateQuery(previousResult: any, { subscriptionData }: any) {
    return {
      systemHealth: subscriptionData.data.systemHealthUpdated,
    }
  },
}

export const updateSystemHealthMutation = gql`
  mutation UpdateSystemHealth($health: UpdateSystemHealthInput!) {
    updateSystemHealth(health: $health)
  }
`

export const getDatabaseOptionsQuery = gql`
  query DatabaseOptions {
    allMolecularDBs {
      id
      name
      version
      archived
      group {
        id
        shortName
      }
    }
  }
`
