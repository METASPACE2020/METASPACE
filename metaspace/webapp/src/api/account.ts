const fetchPostJson = async (url: string, body: object) =>
  await fetch(url, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify(body),
  });

export const signInByEmail = async (email: string, password: string): Promise<boolean> => {
  const response = await fetchPostJson('/auth/signin', { email, password });
  return response.status >= 200 && response.status < 300;
};

export const createAccountByEmail = async (email: string, password: string, name: string) => {
  const response = await fetchPostJson('/auth/createaccount', { email, password, name });
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
};

export const sendPasswordResetToken = async (email: string) => {
  const response = await fetchPostJson('/auth/sendpasswordresettoken', { email });
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
};

export const validatePasswordResetToken = async (token: string, email: string): Promise<boolean> => {
  const response = await fetchPostJson('/auth/validatepasswordresettoken', { token, email });
  return response.status === 200;
};


export const resetPassword = async (token: string, email: string, password: string) => {
  const response = await fetchPostJson('/auth/sendpasswordreset', { token, email, password });
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
};
