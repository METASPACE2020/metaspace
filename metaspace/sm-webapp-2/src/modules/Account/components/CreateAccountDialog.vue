<template>
  <el-dialog
    class="max-w-2xl"
    :title="hasSucceeded ? 'Welcome to METASPACE!' : 'Create an account'"
    v-model="model.dialogVisible"
    :lock-scroll="false"
    :close-on-click-modal="!hasSucceeded"
    @close="onClose"
  >
    <div v-if="hasSucceeded">
      <p>Please click the link in the email we sent you to finish setting up your account.</p>
    </div>
    <div v-else>
      <div class="container">
        <div class="left-side">
          <el-form ref="form" :model="model" :rules="rules">
            <el-row>
              <el-col :span="11">
                <el-form-item prop="firstName">
                  <el-input v-model="model.firstName" type="text" placeholder="First name" required />
                </el-form-item>
              </el-col>
              <el-col :span="2"> &ensp;<!-- el-cols collapse if empty - use a space to prevent this --> </el-col>
              <el-col :span="11">
                <el-form-item prop="lastName">
                  <el-input v-model="model.lastName" type="text" placeholder="Last name" required />
                </el-form-item>
              </el-col>
            </el-row>
            <el-form-item prop="email">
              <el-input v-model="model.email" type="text" placeholder="Email address" required />
            </el-form-item>
            <el-form-item prop="password">
              <el-input
                v-model="model.password"
                type="password"
                placeholder="Password"
                required
                @keypress.enter="onSubmit"
              />
            </el-form-item>
            <el-button data-testid="submit-btn" type="primary" :loading="isSubmitting" @click="onSubmit">
              Create account
            </el-button>
          </el-form>
        </div>
        <div class="mid">
          <div class="divider" />
          <div class="text">or</div>
          <div class="divider" />
        </div>
        <div class="right-side">
          <a class="google-button" href="/api_auth/google" @click="setReturnUrl">
            <google-button>Sign up with Google</google-button>
          </a>
          <ul style="padding: 0 20px">
            <li>Quick sign in with no password</li>
            <li>We only see your name and email address</li>
          </ul>
        </div>
      </div>
      <p style="margin-bottom: 0">
        Already registered? <inter-dialog-link dialog="signIn"> Sign in </inter-dialog-link>
      </p>
    </div>
  </el-dialog>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue'
import { FormInstance } from 'element-plus'
import GoogleButton from './GoogleButton.vue'
import InterDialogLink from './InterDialogLink'
import { createAccountByEmail } from '../../../api/auth'
import reportError from '../../../lib/reportError'
import emailRegex from '../../../lib/emailRegex'
import { setSignInReturnUrl } from '../signInReturnUrl'
import { useStore } from 'vuex'
import { useRoute } from 'vue-router'

export default defineComponent({
  components: {
    GoogleButton,
    InterDialogLink,
  },
  setup() {
    const store = useStore()
    const route = useRoute()
    const isSubmitting = ref(false)
    const hasSucceeded = ref(false)
    const model = ref({
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      dialogVisible: true,
    })

    const rules = {
      firstName: [{ required: true, min: 2, max: 50, message: 'First name is required' }],
      lastName: [{ required: true, min: 2, max: 50, message: 'Last name is required' }],
      email: [
        { required: true, message: 'Email address is required' },
        { pattern: emailRegex, message: 'Must be a valid email address' },
      ],
      password: [{ required: true, min: 8, message: 'Password must be at least 8 characters' }],
    }

    const form = ref<FormInstance>()

    const onSubmit = async () => {
      try {
        await form.value?.validate()
      } catch (err) {
        return
      }
      const { email, password, firstName, lastName } = model.value
      isSubmitting.value = true
      try {
        await createAccountByEmail(email, password, `${firstName} ${lastName}`)
        hasSucceeded.value = true
      } catch (err) {
        reportError(err)
      } finally {
        isSubmitting.value = false
      }
    }

    const setReturnUrl = () => {
      setSignInReturnUrl(route)
    }

    const onClose = () => {
      store.commit('account/hideDialog', 'createAccount')
    }

    return { isSubmitting, hasSucceeded, model, rules, onSubmit, setReturnUrl, onClose, form }
  },
})
</script>

<style scoped lang="scss">
@import 'element-plus/theme-chalk/src/mixins/mixins';

.el-dialog__wrapper ::v-deep(.el-dialog) {
  width: 400px;
}

.container {
  display: flex;
  flex-direction: column;
  align-items: stretch;
}

.left-side {
  flex: 0 1 auto;
}

.right-side {
  flex: 0 1 auto;
}

.mid {
  flex: 0 0 50px;
  display: flex;
  flex-direction: row;
  align-items: center;

  .divider {
    flex-grow: 1;
    border-top: 1px solid lightgrey;
  }
  .text {
    flex: none;
    padding: 0.5em;
  }
}

@include res('sm') {
  .el-dialog__wrapper ::v-deep(.el-dialog) {
    width: 690px;
  }
  .container {
    flex-direction: row;
  }
  .left-side {
    width: 360px;
  }
  .right-side {
    width: 240px;
  }
  .mid {
    flex-direction: column;
    .divider {
      border-top: none;
      border-left: 1px solid lightgrey;
    }
  }
}

.el-button {
  width: 100%;
}

.google-button {
  text-decoration: none;
}
</style>
