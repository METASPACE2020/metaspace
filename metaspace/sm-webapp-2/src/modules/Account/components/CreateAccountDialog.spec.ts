import {flushPromises, mount} from '@vue/test-utils';
import { createRouter, createWebHistory } from 'vue-router';
import { createStore } from 'vuex';
import { beforeEach, afterEach, vi } from 'vitest';
import ElementPlus from 'element-plus';
import CreateAccountDialog from './CreateAccountDialog.vue';
import account from '../store/account';
import { nextTick } from 'vue';
import * as authApi from '../../../api/auth';

vi.mock('../../../api/auth');

// Create a router instance
const localRouter = createRouter({
  history: createWebHistory(),
  routes: [], // Define your routes here
});

const setFormField = (wrapper, fieldName, value) => {
  const formItem = wrapper.findAllComponents({ name: 'ElFormItem' }).find(w => w.props().prop === fieldName);
  formItem.find('input').setValue(value);
};

describe('CreateAccountDialog', () => {
  let store;

  beforeEach(() => {
    vi.clearAllMocks();
    store = createStore({
      modules: {
        account: account,
      },
    });
  });

  afterEach(() => {
    // Clean up or restore any global changes here
    // For example, if you had mocked global functions or properties
  });

  it('should match snapshot', async ({expect}) => {
    const wrapper = mount(CreateAccountDialog, {
      global: {
        plugins: [store, localRouter, ElementPlus],
      },
    });
    await nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should be able to submit a valid form', async ({expect}) => {
    // Import the mocked module

    // Arrange
    const firstName = 'foo';
    const lastName = 'bar';
    const email = 'test@example.com';
    const password = 'abcd1234';
    const wrapper = mount(CreateAccountDialog, {
      global: {
        plugins: [store, localRouter, ElementPlus],
      },
    });
    await nextTick();

    // Act
    setFormField(wrapper, 'firstName', firstName);
    setFormField(wrapper, 'lastName', lastName);
    setFormField(wrapper, 'email', email);
    setFormField(wrapper, 'password', password);
    await wrapper.vm.form?.validate();

    await wrapper.find('[data-testid="submit-btn"]').trigger('click');
    await flushPromises()
    await nextTick();

    // Assert
    expect(authApi.createAccountByEmail).toHaveBeenCalledTimes(1);
    const paragraph = wrapper.find('p');
    expect(paragraph.text()).toContain('Please click the link');
  });

  it('should not submit an invalid form', async ({expect}) => {
    // Arrange
    const wrapper = mount(CreateAccountDialog, {
      global: {
        plugins: [store, localRouter, ElementPlus],
      },
    });
    await nextTick();

    // Act
    setFormField(wrapper, 'firstName', 'foo');
    setFormField(wrapper, 'lastName', 'bar');
    setFormField(wrapper, 'email', 'test@email.com');
    setFormField(wrapper, 'password', ''); // Intentionally left empty
    await wrapper.vm.form?.validate();
    await nextTick();

    await wrapper.find('[data-testid="submit-btn"]').trigger('click');

    // await flushPromises()
    await nextTick();

    // Assert
    expect(authApi.createAccountByEmail).not.toBeCalled();
  });
});
