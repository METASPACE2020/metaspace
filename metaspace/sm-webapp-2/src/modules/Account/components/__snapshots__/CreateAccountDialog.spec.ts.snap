// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateAccountDialog > should match snapshot 1`] = `
"<div data-v-b2005c23=\\"\\" class=\\"mock-el-dialog max-w-2xl\\" title=\\"Create an account\\" modelvalue=\\"true\\" lock-scroll=\\"false\\" close-on-click-modal=\\"true\\">
  <div data-v-b2005c23=\\"\\">
    <div data-v-b2005c23=\\"\\" class=\\"container\\">
      <div data-v-b2005c23=\\"\\" class=\\"left-side\\">
        <form data-v-b2005c23=\\"\\" class=\\"el-form el-form--default el-form--label-right\\">
          <div data-v-b2005c23=\\"\\" class=\\"el-row\\">
            <div data-v-b2005c23=\\"\\" class=\\"el-col el-col-11\\">
              <div data-v-b2005c23=\\"\\" class=\\"el-form-item is-required asterisk-left\\">
                <!--v-if-->
                <div class=\\"el-form-item__content\\">
                  <div data-v-b2005c23=\\"\\" class=\\"el-input\\">
                    <!-- input -->
                    <!-- prepend slot -->
                    <!--v-if-->
                    <div class=\\"el-input__wrapper\\" tabindex=\\"-1\\">
                      <!-- prefix slot -->
                      <!--v-if--><input class=\\"el-input__inner\\" required=\\"\\" type=\\"text\\" autocomplete=\\"off\\" tabindex=\\"0\\" placeholder=\\"First name\\" id=\\"el-id-replaced\\"><!-- suffix slot -->
                      <!--v-if-->
                    </div><!-- append slot -->
                    <!--v-if-->
                  </div>
                </div>
              </div>
            </div>
            <div data-v-b2005c23=\\"\\" class=\\"el-col el-col-2\\"> â€‚
              <!-- el-cols collapse if empty - use a space to prevent this -->
            </div>
            <div data-v-b2005c23=\\"\\" class=\\"el-col el-col-11\\">
              <div data-v-b2005c23=\\"\\" class=\\"el-form-item is-required asterisk-left\\">
                <!--v-if-->
                <div class=\\"el-form-item__content\\">
                  <div data-v-b2005c23=\\"\\" class=\\"el-input\\">
                    <!-- input -->
                    <!-- prepend slot -->
                    <!--v-if-->
                    <div class=\\"el-input__wrapper\\" tabindex=\\"-1\\">
                      <!-- prefix slot -->
                      <!--v-if--><input class=\\"el-input__inner\\" required=\\"\\" type=\\"text\\" autocomplete=\\"off\\" tabindex=\\"0\\" placeholder=\\"Last name\\" id=\\"el-id-replaced\\"><!-- suffix slot -->
                      <!--v-if-->
                    </div><!-- append slot -->
                    <!--v-if-->
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div data-v-b2005c23=\\"\\" class=\\"el-form-item is-required asterisk-left\\">
            <!--v-if-->
            <div class=\\"el-form-item__content\\">
              <div data-v-b2005c23=\\"\\" class=\\"el-input\\">
                <!-- input -->
                <!-- prepend slot -->
                <!--v-if-->
                <div class=\\"el-input__wrapper\\" tabindex=\\"-1\\">
                  <!-- prefix slot -->
                  <!--v-if--><input class=\\"el-input__inner\\" required=\\"\\" type=\\"text\\" autocomplete=\\"off\\" tabindex=\\"0\\" placeholder=\\"Email address\\" id=\\"el-id-replaced\\"><!-- suffix slot -->
                  <!--v-if-->
                </div><!-- append slot -->
                <!--v-if-->
              </div>
            </div>
          </div>
          <div data-v-b2005c23=\\"\\" class=\\"el-form-item is-required asterisk-left\\">
            <!--v-if-->
            <div class=\\"el-form-item__content\\">
              <div data-v-b2005c23=\\"\\" class=\\"el-input\\">
                <!-- input -->
                <!-- prepend slot -->
                <!--v-if-->
                <div class=\\"el-input__wrapper\\" tabindex=\\"-1\\">
                  <!-- prefix slot -->
                  <!--v-if--><input class=\\"el-input__inner\\" required=\\"\\" type=\\"password\\" autocomplete=\\"off\\" tabindex=\\"0\\" placeholder=\\"Password\\" id=\\"el-id-replaced\\"><!-- suffix slot -->
                  <!--v-if-->
                </div><!-- append slot -->
                <!--v-if-->
              </div>
            </div>
          </div><button data-v-b2005c23=\\"\\" ariadisabled=\\"false\\" type=\\"button\\" class=\\"el-button el-button--primary\\" data-testid=\\"submit-btn\\">
            <!--v-if--><span class=\\"\\"> Create account </span>
          </button>
        </form>
      </div>
      <div data-v-b2005c23=\\"\\" class=\\"mid\\">
        <div data-v-b2005c23=\\"\\" class=\\"divider\\"></div>
        <div data-v-b2005c23=\\"\\" class=\\"text\\">or</div>
        <div data-v-b2005c23=\\"\\" class=\\"divider\\"></div>
      </div>
      <div data-v-b2005c23=\\"\\" class=\\"right-side\\"><a data-v-b2005c23=\\"\\" class=\\"google-button\\" href=\\"/api_auth/google\\">
          <div data-v-b2005c23=\\"\\" class=\\"abcRioButton abcRioButtonBlue\\">
            <div class=\\"abcRioButtonContentWrapper\\">
              <div class=\\"abcRioButtonIcon\\" style=\\"padding:10px;\\">
                <div style=\\"width:18px;height:18px;\\" class=\\"abcRioButtonSvgImageWithFallback abcRioButtonIconImage abcRioButtonIconImage18\\"><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18px\\" height=\\"18px\\" viewBox=\\"0 0 48 48\\" class=\\"abcRioButtonSvg\\">
                    <g>
                      <path fill=\\"#EA4335\\" d=\\"M24 9.5c3.54 0 6.71 1.22 9.21 3.6l6.85-6.85C35.9 2.38 30.47 0 24 0 14.62 0 6.51 5.38 2.56 13.22l7.98 6.19C12.43 13.72 17.74 9.5 24 9.5z\\"></path>
                      <path fill=\\"#4285F4\\" d=\\"M46.98 24.55c0-1.57-.15-3.09-.38-4.55H24v9.02h12.94c-.58 2.96-2.26 5.48-4.78 7.18l7.73 6c4.51-4.18 7.09-10.36 7.09-17.65z\\"></path>
                      <path fill=\\"#FBBC05\\" d=\\"M10.53 28.59c-.48-1.45-.76-2.99-.76-4.59s.27-3.14.76-4.59l-7.98-6.19C.92 16.46 0 20.12 0 24c0 3.88.92 7.54 2.56 10.78l7.97-6.19z\\"></path>
                      <path fill=\\"#34A853\\" d=\\"M24 48c6.48 0 11.93-2.13 15.89-5.81l-7.73-6c-2.15 1.45-4.92 2.3-8.16 2.3-6.26 0-11.57-4.22-13.47-9.91l-7.98 6.19C6.51 42.62 14.62 48 24 48z\\"></path>
                      <path fill=\\"none\\" d=\\"M0 0h48v48H0z\\"></path>
                    </g>
                  </svg></div>
              </div><span style=\\"font-size: 16px; line-height: 38px;\\" class=\\"abcRioButtonContents\\">Sign up with Google</span>
            </div>
          </div>
        </a>
        <ul data-v-b2005c23=\\"\\" style=\\"padding: 0px 20px;\\">
          <li data-v-b2005c23=\\"\\">Quick sign in with no password</li>
          <li data-v-b2005c23=\\"\\">We only see your name and email address</li>
        </ul>
      </div>
    </div>
    <p data-v-b2005c23=\\"\\" style=\\"margin-bottom: 0px;\\"> Already registered? <a data-v-b2005c23=\\"\\" href=\\"/account/sign-in\\"> Sign in </a></p>
  </div>
</div>"
`;
